std/utils.asm

```diff
@@ -0,0 +1,31 @@
+
+/// Evaluates to folder(...folder(folder(initial, f(0)), f(1)) ..., f(length - 1)),
+/// i.e. calls f(0), f(1), ..., f(length - 1) and combines the results
+/// using the function `folder`, starting with the value `initial`.
+///
+/// See `sum` for an example use.
+let fold = |length, f, initial, folder|
+    if length <= 0 {
+        initial
+    } else {
+        folder(fold((length - 1), f, initial, folder), f((length - 1)))
+    };
+
+/// Evaluates to the array [f(0), f(1), ..., f(length - 1)].
+let make_array = |length, f| fold(length, f, [], |acc, e| (acc + [e]));
+
+/// Evaluates to the array [f(arr[0]), f(arr[1]), ..., f(arr[length - 1])].
+let map = |arr, length, f| make_array(length, |i| f(arr[i]));
```

Comment:
Would it be possible to get the length from the input array?

Reply:
I think not currently, @chriseth was planning `arr.length` at some point in the past

Reply:
Yes, we have to come up with a convention. `x.y` currently is not yet parsed and is also much more difficult to resolve.

Reply:
... and there is no way to define such things in user code, so I'm reluctant to add it as a feature for built-in types.
We could do something like `use std::builtins::len` and then you can use the `len` function on arrays.